{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KLHL3</td>\n",
       "      <td>5</td>\n",
       "      <td>136973018</td>\n",
       "      <td>136973018</td>\n",
       "      <td>TCGA-OR-A5J1-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KLHDC4</td>\n",
       "      <td>16</td>\n",
       "      <td>87742934</td>\n",
       "      <td>87742934</td>\n",
       "      <td>TCGA-OR-A5J1-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIST1H3B</td>\n",
       "      <td>6</td>\n",
       "      <td>26031919</td>\n",
       "      <td>26031919</td>\n",
       "      <td>TCGA-OR-A5J1-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GABRA1</td>\n",
       "      <td>5</td>\n",
       "      <td>161300197</td>\n",
       "      <td>161300197</td>\n",
       "      <td>TCGA-OR-A5J1-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCARF1</td>\n",
       "      <td>17</td>\n",
       "      <td>1543262</td>\n",
       "      <td>1543262</td>\n",
       "      <td>TCGA-OR-A5J1-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Chromosome  Start_position  End_position  \\\n",
       "0       KLHL3          5       136973018     136973018   \n",
       "1      KLHDC4         16        87742934      87742934   \n",
       "2    HIST1H3B          6        26031919      26031919   \n",
       "3      GABRA1          5       161300197     161300197   \n",
       "4      SCARF1         17         1543262       1543262   \n",
       "\n",
       "           Tumor_Sample_Barcode  \n",
       "0  TCGA-OR-A5J1-01A-11D-A29I-10  \n",
       "1  TCGA-OR-A5J1-01A-11D-A29I-10  \n",
       "2  TCGA-OR-A5J1-01A-11D-A29I-10  \n",
       "3  TCGA-OR-A5J1-01A-11D-A29I-10  \n",
       "4  TCGA-OR-A5J1-01A-11D-A29I-10  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pylab import *\n",
    "\n",
    "FILE = \"ACC.txt\"\n",
    "\n",
    "maf = pd.read_table(FILE, skiprows = 4, usecols=[\"Hugo_Symbol\", \"Chromosome\", \"Start_position\", \"End_position\", \"Tumor_Sample_Barcode\"])\n",
    "maf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13129\n"
     ]
    }
   ],
   "source": [
    "print len(maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>miR Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>ID=MI0022705;Alias=MI0022705;Name=hsa-mir-6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17409</td>\n",
       "      <td>17431</td>\n",
       "      <td>ID=MIMAT0027618;Alias=MIMAT0027618;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17369</td>\n",
       "      <td>17391</td>\n",
       "      <td>ID=MIMAT0027619;Alias=MIMAT0027619;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>ID=MI0006363;Alias=MI0006363;Name=hsa-mir-1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>30438</td>\n",
       "      <td>30458</td>\n",
       "      <td>ID=MIMAT0005890;Alias=MIMAT0005890;Name=hsa-mi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome                      Type  Start    End  \\\n",
       "0       chr1  miRNA_primary_transcript  17369  17436   \n",
       "1       chr1                     miRNA  17409  17431   \n",
       "2       chr1                     miRNA  17369  17391   \n",
       "3       chr1  miRNA_primary_transcript  30366  30503   \n",
       "4       chr1                     miRNA  30438  30458   \n",
       "\n",
       "                                            miR Name  \n",
       "0   ID=MI0022705;Alias=MI0022705;Name=hsa-mir-6859-1  \n",
       "1  ID=MIMAT0027618;Alias=MIMAT0027618;Name=hsa-mi...  \n",
       "2  ID=MIMAT0027619;Alias=MIMAT0027619;Name=hsa-mi...  \n",
       "3   ID=MI0006363;Alias=MI0006363;Name=hsa-mir-1302-2  \n",
       "4  ID=MIMAT0005890;Alias=MIMAT0005890;Name=hsa-mi...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FILE = \"../hw6/hsa.gff3\"\n",
    "df = pd.read_table(FILE, comment=\"#\", usecols=[\"Chromosome\", \"Start\", \"End\", \"Type\", \"miR Name\"])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\thas 398\tmiRNA records\n",
      "chr2\thas 271\tmiRNA records\n",
      "chr3\thas 227\tmiRNA records\n",
      "chr4\thas 144\tmiRNA records\n",
      "chr5\thas 175\tmiRNA records\n",
      "chr6\thas 162\tmiRNA records\n",
      "chr7\thas 198\tmiRNA records\n",
      "chr8\thas 224\tmiRNA records\n",
      "chr9\thas 207\tmiRNA records\n",
      "chr10\thas 167\tmiRNA records\n",
      "chr11\thas 255\tmiRNA records\n",
      "chr12\thas 186\tmiRNA records\n",
      "chr13\thas 97\tmiRNA records\n",
      "chr14\thas 258\tmiRNA records\n",
      "chr15\thas 158\tmiRNA records\n",
      "chr16\thas 194\tmiRNA records\n",
      "chr17\thas 284\tmiRNA records\n",
      "chr18\thas 83\tmiRNA records\n",
      "chr19\thas 376\tmiRNA records\n",
      "chr20\thas 115\tmiRNA records\n",
      "chr21\thas 54\tmiRNA records\n",
      "chr22\thas 118\tmiRNA records\n",
      "chrX\thas 310\tmiRNA records\n",
      "chrY\thas 4\tmiRNA records\n"
     ]
    }
   ],
   "source": [
    "chr2df = {}\n",
    "for chr in [str(i) for i in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chr2df[chr] = df[df[\"Chromosome\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d\\tmiRNA records\" % (chr, len(chr2df[chr]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def pos2miR(chr, start, end):\n",
    "    chunk = chr2df[str(chr)]  # shrink the search space to the given chr\n",
    "    out = []\n",
    "    for idx, row in chunk.iterrows():\n",
    "        if row[\"Start\"] <= start <= row[\"End\"] or \\\n",
    "           row[\"Start\"] <=  end  <= row[\"End\"] or \\\n",
    "           (start < row[\"Start\"] and row[\"End\"] < end):\n",
    "            # get the name (either hsa-mir-### or hsa-miR-###)\n",
    "            miR = row[\"Type\"] + \";\" + str(row[\"Start\"]) + \";\" + str(row[\"End\"]) + \";\" + row[\"miR Name\"].split(\"Name=\")[1].split(\";\")[0]\n",
    "            out.append(miR)\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['miRNA_primary_transcript;17369;17436;hsa-mir-6859-1',\n",
       " 'miRNA;17369;17391;hsa-miR-6859-3p']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test\n",
    "pos2miR(1, 17370, 17370)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['miRNA_primary_transcript;2477232;2477295;hsa-mir-6089-2',\n",
       " 'miRNA;2477270;2477293;hsa-miR-6089']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test2\n",
    "pos2miR(\"Y\", 2477271, 2477370)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MIR4732 17 27188561 27188561 TCGA-OR-A5J5-01A-11D-A29I-10 hits on miRNA_primary_transcript;27188551;27188636;hsa-mir-144\n",
      "MIR34C 11 111384236 111384236 TCGA-OR-A5J8-01A-11D-A29I-10 hits on miRNA_primary_transcript;111384164;111384240;hsa-mir-34c\n",
      "MIR506 X 146312342 146312342 TCGA-OR-A5JB-01A-11D-A29I-10 hits on miRNA_primary_transcript;146312238;146312361;hsa-mir-506\n",
      "MIR520C 19 54210730 54210730 TCGA-OR-A5JC-01A-11D-A29I-10 hits on miRNA_primary_transcript;54210707;54210793;hsa-mir-520c\n",
      "MIR520C 19 54210730 54210730 TCGA-OR-A5JC-01A-11D-A29I-10 hits on miRNA;54210722;54210743;hsa-miR-520c-5p\n",
      "RP5-963E22.4 20 61809858 61809858 TCGA-OR-A5JW-01A-11D-A29I-10 hits on miRNA_primary_transcript;61809852;61809938;hsa-mir-124-3\n",
      "MIR192 11 64658676 64658676 TCGA-OR-A5K4-01A-11D-A29I-10 hits on miRNA_primary_transcript;64658609;64658718;hsa-mir-192\n",
      "MIR192 11 64658676 64658676 TCGA-OR-A5K4-01A-11D-A29I-10 hits on miRNA;64658675;64658695;hsa-miR-192-5p\n",
      "MIR509-3 X 146341225 146341225 TCGA-OR-A5KB-01A-11D-A30A-10 hits on miRNA_primary_transcript;146341170;146341244;hsa-mir-509-3\n",
      "MIR509-3 X 146341225 146341225 TCGA-OR-A5KB-01A-11D-A30A-10 hits on miRNA;146341214;146341235;hsa-miR-509-3-5p\n",
      "RTL1 14 101351118 101351118 TCGA-OR-A5KB-01A-11D-A30A-10 hits on miRNA_primary_transcript;101351039;101351120;hsa-mir-136\n",
      "MIR506 X 146312360 146312360 TCGA-OR-A5KB-01A-11D-A30A-10 hits on miRNA_primary_transcript;146312238;146312361;hsa-mir-506\n",
      "MIR892B X 145078780 145078780 TCGA-OR-A5KB-01A-11D-A30A-10 hits on miRNA_primary_transcript;145078716;145078792;hsa-mir-892b\n",
      "MIR517A 19 54215536 54215536 TCGA-OR-A5KB-01A-11D-A30A-10 hits on miRNA_primary_transcript;54215522;54215608;hsa-mir-517a\n",
      "MIR517A 19 54215536 54215536 TCGA-OR-A5KB-01A-11D-A30A-10 hits on miRNA;54215536;54215557;hsa-miR-517-5p\n",
      "MIR512-1 19 54169970 54169970 TCGA-OR-A5KT-01A-11D-A29I-10 hits on miRNA_primary_transcript;54169933;54170016;hsa-mir-512-1\n",
      "MIR520F 19 54185480 54185480 TCGA-OR-A5KZ-01A-11D-A29I-10 hits on miRNA_primary_transcript;54185413;54185499;hsa-mir-520f\n",
      "MIR520F 19 54185480 54185480 TCGA-OR-A5KZ-01A-11D-A29I-10 hits on miRNA;54185467;54185488;hsa-miR-520f-3p\n",
      "MIR298 20 57393302 57393302 TCGA-OR-A5LB-01A-11D-A29I-10 hits on miRNA_primary_transcript;57393281;57393368;hsa-mir-298\n",
      "MIR512-2 19 54172505 54172505 TCGA-OR-A5LD-01A-11D-A29I-10 hits on miRNA_primary_transcript;54172411;54172508;hsa-mir-512-2\n",
      "MIR542 X 133675462 133675462 TCGA-OR-A5LI-01A-11D-A30A-10 hits on miRNA_primary_transcript;133675371;133675467;hsa-mir-542\n"
     ]
    }
   ],
   "source": [
    "c = pd.DataFrame()\n",
    "ms = pd.DataFrame()\n",
    "me = pd.DataFrame()\n",
    "rs = pd.DataFrame()\n",
    "re = pd.DataFrame()\n",
    "t = pd.DataFrame()\n",
    "i = pd.DataFrame()\n",
    "s = pd.DataFrame()\n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    if row[\"Chromosome\"] == \"MT\":\n",
    "        continue\n",
    "    hits = pos2miR(row[\"Chromosome\"], (row[\"Start_position\"]), (row[\"End_position\"]))\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_position\"]])\n",
    "        me = me.append([row[\"End_position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "miRNA;146341214;146341235;hsa-miR-509-3-5p                    1\n",
       "miRNA;54185467;54185488;hsa-miR-520f-3p                       1\n",
       "miRNA;54210722;54210743;hsa-miR-520c-5p                       1\n",
       "miRNA;54215536;54215557;hsa-miR-517-5p                        1\n",
       "miRNA;64658675;64658695;hsa-miR-192-5p                        1\n",
       "miRNA_primary_transcript;101351039;101351120;hsa-mir-136      1\n",
       "miRNA_primary_transcript;111384164;111384240;hsa-mir-34c      1\n",
       "miRNA_primary_transcript;133675371;133675467;hsa-mir-542      1\n",
       "miRNA_primary_transcript;145078716;145078792;hsa-mir-892b     1\n",
       "miRNA_primary_transcript;146312238;146312361;hsa-mir-506      2\n",
       "miRNA_primary_transcript;146341170;146341244;hsa-mir-509-3    1\n",
       "miRNA_primary_transcript;27188551;27188636;hsa-mir-144        1\n",
       "miRNA_primary_transcript;54169933;54170016;hsa-mir-512-1      1\n",
       "miRNA_primary_transcript;54172411;54172508;hsa-mir-512-2      1\n",
       "miRNA_primary_transcript;54185413;54185499;hsa-mir-520f       1\n",
       "miRNA_primary_transcript;54210707;54210793;hsa-mir-520c       1\n",
       "miRNA_primary_transcript;54215522;54215608;hsa-mir-517a       1\n",
       "miRNA_primary_transcript;57393281;57393368;hsa-mir-298        1\n",
       "miRNA_primary_transcript;61809852;61809938;hsa-mir-124-3      1\n",
       "miRNA_primary_transcript;64658609;64658718;hsa-mir-192        1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.Series(miR2hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TABLE2 = \"C:/Users/jtso/Downloads/targetScanS.gz\"  # miRNA binding sites\n",
    "TABLE3 = \"C:/Users/jtso/Downloads/wgEncodeGencodePolyaV19 (1).gz\"  # polyA sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>879822</td>\n",
       "      <td>879830</td>\n",
       "      <td>SAMD11:miR-504</td>\n",
       "      <td>90</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>900599</td>\n",
       "      <td>900606</td>\n",
       "      <td>KLHL17:miR-299/299-3p</td>\n",
       "      <td>26</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>900605</td>\n",
       "      <td>900612</td>\n",
       "      <td>KLHL17:miR-124/506</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>900933</td>\n",
       "      <td>900941</td>\n",
       "      <td>KLHL17:miR-19</td>\n",
       "      <td>82</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>901054</td>\n",
       "      <td>901061</td>\n",
       "      <td>KLHL17:miR-137</td>\n",
       "      <td>14</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>592</td>\n",
       "      <td>chr1</td>\n",
       "      <td>990829</td>\n",
       "      <td>990836</td>\n",
       "      <td>AGRN:miR-144</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>592</td>\n",
       "      <td>chr1</td>\n",
       "      <td>990831</td>\n",
       "      <td>990838</td>\n",
       "      <td>AGRN:miR-27ab</td>\n",
       "      <td>47</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>592</td>\n",
       "      <td>chr1</td>\n",
       "      <td>990831</td>\n",
       "      <td>990838</td>\n",
       "      <td>AGRN:miR-128</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>592</td>\n",
       "      <td>chr1</td>\n",
       "      <td>990834</td>\n",
       "      <td>990841</td>\n",
       "      <td>AGRN:miR-224</td>\n",
       "      <td>50</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>593</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1169073</td>\n",
       "      <td>1169080</td>\n",
       "      <td>B3GALT6:miR-340/340-5p</td>\n",
       "      <td>43</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>594</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1270756</td>\n",
       "      <td>1270763</td>\n",
       "      <td>DVL1:miR-139-5p</td>\n",
       "      <td>55</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1321936</td>\n",
       "      <td>1321944</td>\n",
       "      <td>CCNL2:miR-28/28-5p/708</td>\n",
       "      <td>69</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1321941</td>\n",
       "      <td>1321948</td>\n",
       "      <td>CCNL2:miR-138</td>\n",
       "      <td>32</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1321986</td>\n",
       "      <td>1321993</td>\n",
       "      <td>CCNL2:miR-421</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1322008</td>\n",
       "      <td>1322015</td>\n",
       "      <td>CCNL2:miR-23ab</td>\n",
       "      <td>51</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1322057</td>\n",
       "      <td>1322064</td>\n",
       "      <td>CCNL2:miR-26ab/1297:2</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1322276</td>\n",
       "      <td>1322284</td>\n",
       "      <td>CCNL2:miR-29abc</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1326479</td>\n",
       "      <td>1326486</td>\n",
       "      <td>CCNL2:miR-320/320abcd</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1326654</td>\n",
       "      <td>1326661</td>\n",
       "      <td>CCNL2:miR-130/301</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1326964</td>\n",
       "      <td>1326971</td>\n",
       "      <td>CCNL2:miR-543</td>\n",
       "      <td>57</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1326965</td>\n",
       "      <td>1326972</td>\n",
       "      <td>CCNL2:miR-181</td>\n",
       "      <td>51</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1327888</td>\n",
       "      <td>1327896</td>\n",
       "      <td>CCNL2:miR-217</td>\n",
       "      <td>93</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1327934</td>\n",
       "      <td>1327941</td>\n",
       "      <td>CCNL2:miR-26ab/1297:1</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>595</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1328082</td>\n",
       "      <td>1328089</td>\n",
       "      <td>CCNL2:miR-185/882</td>\n",
       "      <td>18</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>597</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1593136</td>\n",
       "      <td>1593143</td>\n",
       "      <td>RP11-345P4.4:miR-200bc/429</td>\n",
       "      <td>90</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>597</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1594097</td>\n",
       "      <td>1594105</td>\n",
       "      <td>RP11-345P4.4:miR-155</td>\n",
       "      <td>65</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>597</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1595719</td>\n",
       "      <td>1595727</td>\n",
       "      <td>RP11-345P4.4:miR-137</td>\n",
       "      <td>37</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>597</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1682693</td>\n",
       "      <td>1682701</td>\n",
       "      <td>NADK:miR-488</td>\n",
       "      <td>98</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>597</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1682759</td>\n",
       "      <td>1682767</td>\n",
       "      <td>NADK:miR-494</td>\n",
       "      <td>99</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>597</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1683543</td>\n",
       "      <td>1683551</td>\n",
       "      <td>NADK:miR-30a/30a-5p/30b/30b-5p/30cde/384-5p</td>\n",
       "      <td>79</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54169</th>\n",
       "      <td>971</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50691444</td>\n",
       "      <td>50691452</td>\n",
       "      <td>MAPK12:miR-125/351</td>\n",
       "      <td>94</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54170</th>\n",
       "      <td>971</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50702382</td>\n",
       "      <td>50702389</td>\n",
       "      <td>MAPK11:let-7/98</td>\n",
       "      <td>32</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54171</th>\n",
       "      <td>971</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50713669</td>\n",
       "      <td>50713677</td>\n",
       "      <td>PLXNB2:miR-126/126-3p</td>\n",
       "      <td>94</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54172</th>\n",
       "      <td>971</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50713793</td>\n",
       "      <td>50713801</td>\n",
       "      <td>PLXNB2:miR-138:3</td>\n",
       "      <td>44</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54173</th>\n",
       "      <td>971</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50713843</td>\n",
       "      <td>50713851</td>\n",
       "      <td>PLXNB2:miR-124/506</td>\n",
       "      <td>95</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54174</th>\n",
       "      <td>971</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50713924</td>\n",
       "      <td>50713931</td>\n",
       "      <td>PLXNB2:miR-138:2</td>\n",
       "      <td>44</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54175</th>\n",
       "      <td>971</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50713929</td>\n",
       "      <td>50713937</td>\n",
       "      <td>PLXNB2:miR-138:1</td>\n",
       "      <td>44</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54176</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50882733</td>\n",
       "      <td>50882740</td>\n",
       "      <td>SAPS2:miR-196ab</td>\n",
       "      <td>79</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54177</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50882787</td>\n",
       "      <td>50882795</td>\n",
       "      <td>SAPS2:miR-103/107</td>\n",
       "      <td>99</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54178</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50882927</td>\n",
       "      <td>50882935</td>\n",
       "      <td>SAPS2:miR-137</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54179</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50883100</td>\n",
       "      <td>50883107</td>\n",
       "      <td>SAPS2:miR-19</td>\n",
       "      <td>66</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54180</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50883101</td>\n",
       "      <td>50883108</td>\n",
       "      <td>SAPS2:miR-130/301</td>\n",
       "      <td>87</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54181</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50883103</td>\n",
       "      <td>50883110</td>\n",
       "      <td>SAPS2:miR-17-5p/20/93.mr/106/519.d</td>\n",
       "      <td>78</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54182</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50883244</td>\n",
       "      <td>50883251</td>\n",
       "      <td>SAPS2:miR-124/506</td>\n",
       "      <td>46</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54183</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50885466</td>\n",
       "      <td>50885473</td>\n",
       "      <td>SBF1:miR-200bc/429</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54184</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50885487</td>\n",
       "      <td>50885495</td>\n",
       "      <td>SBF1:miR-30a/30a-5p/30b/30b-5p/30cde/384-5p</td>\n",
       "      <td>51</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54185</th>\n",
       "      <td>973</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50941729</td>\n",
       "      <td>50941737</td>\n",
       "      <td>LMF2:miR-124/506</td>\n",
       "      <td>89</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54186</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51169809</td>\n",
       "      <td>51169816</td>\n",
       "      <td>SHANK3:miR-615-3p</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54187</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51170262</td>\n",
       "      <td>51170270</td>\n",
       "      <td>SHANK3:miR-7/7ab</td>\n",
       "      <td>78</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54188</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51170282</td>\n",
       "      <td>51170290</td>\n",
       "      <td>SHANK3:miR-34a/34b-5p/34c/34c-5p/449/449abc/699:1</td>\n",
       "      <td>45</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54189</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51170288</td>\n",
       "      <td>51170296</td>\n",
       "      <td>SHANK3:miR-34a/34b-5p/34c/34c-5p/449/449abc/699:2</td>\n",
       "      <td>45</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54190</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171069</td>\n",
       "      <td>51171077</td>\n",
       "      <td>SHANK3:miR-128</td>\n",
       "      <td>69</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54191</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171070</td>\n",
       "      <td>51171077</td>\n",
       "      <td>SHANK3:miR-27ab</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54192</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171451</td>\n",
       "      <td>51171458</td>\n",
       "      <td>SHANK3:miR-10</td>\n",
       "      <td>88</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54193</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171452</td>\n",
       "      <td>51171460</td>\n",
       "      <td>SHANK3:miR-504</td>\n",
       "      <td>99</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54194</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171554</td>\n",
       "      <td>51171561</td>\n",
       "      <td>SHANK3:miR-491/491-5p</td>\n",
       "      <td>67</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54195</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171565</td>\n",
       "      <td>51171572</td>\n",
       "      <td>SHANK3:miR-376c</td>\n",
       "      <td>72</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54196</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171568</td>\n",
       "      <td>51171575</td>\n",
       "      <td>SHANK3:miR-494</td>\n",
       "      <td>83</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54197</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171600</td>\n",
       "      <td>51171608</td>\n",
       "      <td>SHANK3:miR-374/374ab</td>\n",
       "      <td>99</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54198</th>\n",
       "      <td>975</td>\n",
       "      <td>chr22</td>\n",
       "      <td>51171625</td>\n",
       "      <td>51171632</td>\n",
       "      <td>SHANK3:miR-340/340-5p</td>\n",
       "      <td>51</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54199 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #bin  chrom  chromStart  chromEnd  \\\n",
       "0       591   chr1      879822    879830   \n",
       "1       591   chr1      900599    900606   \n",
       "2       591   chr1      900605    900612   \n",
       "3       591   chr1      900933    900941   \n",
       "4       591   chr1      901054    901061   \n",
       "5       592   chr1      990829    990836   \n",
       "6       592   chr1      990831    990838   \n",
       "7       592   chr1      990831    990838   \n",
       "8       592   chr1      990834    990841   \n",
       "9       593   chr1     1169073   1169080   \n",
       "10      594   chr1     1270756   1270763   \n",
       "11      595   chr1     1321936   1321944   \n",
       "12      595   chr1     1321941   1321948   \n",
       "13      595   chr1     1321986   1321993   \n",
       "14      595   chr1     1322008   1322015   \n",
       "15      595   chr1     1322057   1322064   \n",
       "16      595   chr1     1322276   1322284   \n",
       "17      595   chr1     1326479   1326486   \n",
       "18      595   chr1     1326654   1326661   \n",
       "19      595   chr1     1326964   1326971   \n",
       "20      595   chr1     1326965   1326972   \n",
       "21      595   chr1     1327888   1327896   \n",
       "22      595   chr1     1327934   1327941   \n",
       "23      595   chr1     1328082   1328089   \n",
       "24      597   chr1     1593136   1593143   \n",
       "25      597   chr1     1594097   1594105   \n",
       "26      597   chr1     1595719   1595727   \n",
       "27      597   chr1     1682693   1682701   \n",
       "28      597   chr1     1682759   1682767   \n",
       "29      597   chr1     1683543   1683551   \n",
       "...     ...    ...         ...       ...   \n",
       "54169   971  chr22    50691444  50691452   \n",
       "54170   971  chr22    50702382  50702389   \n",
       "54171   971  chr22    50713669  50713677   \n",
       "54172   971  chr22    50713793  50713801   \n",
       "54173   971  chr22    50713843  50713851   \n",
       "54174   971  chr22    50713924  50713931   \n",
       "54175   971  chr22    50713929  50713937   \n",
       "54176   973  chr22    50882733  50882740   \n",
       "54177   973  chr22    50882787  50882795   \n",
       "54178   973  chr22    50882927  50882935   \n",
       "54179   973  chr22    50883100  50883107   \n",
       "54180   973  chr22    50883101  50883108   \n",
       "54181   973  chr22    50883103  50883110   \n",
       "54182   973  chr22    50883244  50883251   \n",
       "54183   973  chr22    50885466  50885473   \n",
       "54184   973  chr22    50885487  50885495   \n",
       "54185   973  chr22    50941729  50941737   \n",
       "54186   975  chr22    51169809  51169816   \n",
       "54187   975  chr22    51170262  51170270   \n",
       "54188   975  chr22    51170282  51170290   \n",
       "54189   975  chr22    51170288  51170296   \n",
       "54190   975  chr22    51171069  51171077   \n",
       "54191   975  chr22    51171070  51171077   \n",
       "54192   975  chr22    51171451  51171458   \n",
       "54193   975  chr22    51171452  51171460   \n",
       "54194   975  chr22    51171554  51171561   \n",
       "54195   975  chr22    51171565  51171572   \n",
       "54196   975  chr22    51171568  51171575   \n",
       "54197   975  chr22    51171600  51171608   \n",
       "54198   975  chr22    51171625  51171632   \n",
       "\n",
       "                                                    name  score strand  \n",
       "0                                         SAMD11:miR-504     90      +  \n",
       "1                                  KLHL17:miR-299/299-3p     26      +  \n",
       "2                                     KLHL17:miR-124/506      7      +  \n",
       "3                                          KLHL17:miR-19     82      +  \n",
       "4                                         KLHL17:miR-137     14      +  \n",
       "5                                           AGRN:miR-144     10      +  \n",
       "6                                          AGRN:miR-27ab     47      +  \n",
       "7                                           AGRN:miR-128     16      +  \n",
       "8                                           AGRN:miR-224     50      +  \n",
       "9                                 B3GALT6:miR-340/340-5p     43      +  \n",
       "10                                       DVL1:miR-139-5p     55      -  \n",
       "11                                CCNL2:miR-28/28-5p/708     69      -  \n",
       "12                                         CCNL2:miR-138     32      -  \n",
       "13                                         CCNL2:miR-421     34      -  \n",
       "14                                        CCNL2:miR-23ab     51      -  \n",
       "15                                 CCNL2:miR-26ab/1297:2     14      -  \n",
       "16                                       CCNL2:miR-29abc      7      -  \n",
       "17                                 CCNL2:miR-320/320abcd     39      -  \n",
       "18                                     CCNL2:miR-130/301     25      -  \n",
       "19                                         CCNL2:miR-543     57      -  \n",
       "20                                         CCNL2:miR-181     51      -  \n",
       "21                                         CCNL2:miR-217     93      -  \n",
       "22                                 CCNL2:miR-26ab/1297:1     14      -  \n",
       "23                                     CCNL2:miR-185/882     18      -  \n",
       "24                            RP11-345P4.4:miR-200bc/429     90      -  \n",
       "25                                  RP11-345P4.4:miR-155     65      -  \n",
       "26                                  RP11-345P4.4:miR-137     37      -  \n",
       "27                                          NADK:miR-488     98      -  \n",
       "28                                          NADK:miR-494     99      -  \n",
       "29           NADK:miR-30a/30a-5p/30b/30b-5p/30cde/384-5p     79      -  \n",
       "...                                                  ...    ...    ...  \n",
       "54169                                 MAPK12:miR-125/351     94      -  \n",
       "54170                                    MAPK11:let-7/98     32      -  \n",
       "54171                              PLXNB2:miR-126/126-3p     94      -  \n",
       "54172                                   PLXNB2:miR-138:3     44      -  \n",
       "54173                                 PLXNB2:miR-124/506     95      -  \n",
       "54174                                   PLXNB2:miR-138:2     44      -  \n",
       "54175                                   PLXNB2:miR-138:1     44      -  \n",
       "54176                                    SAPS2:miR-196ab     79      +  \n",
       "54177                                  SAPS2:miR-103/107     99      +  \n",
       "54178                                      SAPS2:miR-137     41      +  \n",
       "54179                                       SAPS2:miR-19     66      +  \n",
       "54180                                  SAPS2:miR-130/301     87      +  \n",
       "54181                 SAPS2:miR-17-5p/20/93.mr/106/519.d     78      +  \n",
       "54182                                  SAPS2:miR-124/506     46      +  \n",
       "54183                                 SBF1:miR-200bc/429     12      -  \n",
       "54184        SBF1:miR-30a/30a-5p/30b/30b-5p/30cde/384-5p     51      -  \n",
       "54185                                   LMF2:miR-124/506     89      -  \n",
       "54186                                  SHANK3:miR-615-3p      7      +  \n",
       "54187                                   SHANK3:miR-7/7ab     78      +  \n",
       "54188  SHANK3:miR-34a/34b-5p/34c/34c-5p/449/449abc/699:1     45      +  \n",
       "54189  SHANK3:miR-34a/34b-5p/34c/34c-5p/449/449abc/699:2     45      +  \n",
       "54190                                     SHANK3:miR-128     69      +  \n",
       "54191                                    SHANK3:miR-27ab      5      +  \n",
       "54192                                      SHANK3:miR-10     88      +  \n",
       "54193                                     SHANK3:miR-504     99      +  \n",
       "54194                              SHANK3:miR-491/491-5p     67      +  \n",
       "54195                                    SHANK3:miR-376c     72      +  \n",
       "54196                                     SHANK3:miR-494     83      +  \n",
       "54197                               SHANK3:miR-374/374ab     99      +  \n",
       "54198                              SHANK3:miR-340/340-5p     51      +  \n",
       "\n",
       "[54199 rows x 7 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi = pd.read_table(TABLE2)\n",
    "hi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\thas 5456  \t miR genes\n",
      "chr2\thas 4157  \t miR genes\n",
      "chr3\thas 3494  \t miR genes\n",
      "chr4\thas 2383  \t miR genes\n",
      "chr5\thas 3058  \t miR genes\n",
      "chr6\thas 2613  \t miR genes\n",
      "chr7\thas 2554  \t miR genes\n",
      "chr8\thas 2073  \t miR genes\n",
      "chr9\thas 2287  \t miR genes\n",
      "chr10\thas 2302  \t miR genes\n",
      "chr11\thas 2784  \t miR genes\n",
      "chr12\thas 3445  \t miR genes\n",
      "chr13\thas 1084  \t miR genes\n",
      "chr14\thas 1884  \t miR genes\n",
      "chr15\thas 1784  \t miR genes\n",
      "chr16\thas 1836  \t miR genes\n",
      "chr17\thas 3095  \t miR genes\n",
      "chr18\thas 1188  \t miR genes\n",
      "chr19\thas 1463  \t miR genes\n",
      "chr20\thas 1550  \t miR genes\n",
      "chr21\thas 538  \t miR genes\n",
      "chr22\thas 1044  \t miR genes\n",
      "chrX\thas 2080  \t miR genes\n",
      "chrY\thas 47  \t miR genes\n"
     ]
    }
   ],
   "source": [
    "chrom2hi = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2hi[chr] = hi[hi[\"chrom\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d  \\t miR genes\" % (chr, len(chrom2hi[chr]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LPAL2 6 160903734 160903734 TCGA-OR-A5JC-01A-11D-A29I-10 hits on miR binding site;160903734;160903741;LPAL2:miR-376/376ab/376b-3p:2\n"
     ]
    }
   ],
   "source": [
    "def pos2miRbind(chr, start, end):\n",
    "    area = chrom2hi[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"chromStart\"] <= start <= row[\"chromEnd\"] or \\\n",
    "           row[\"chromStart\"] <= end <= row[\"chromEnd\"] or \\\n",
    "           (start < row[\"chromStart\"] and row[\"chromEnd\"] < end):\n",
    "            miR = \"miR binding site\" + \";\" + str(row[\"chromStart\"]) + \";\" + str(row[\"chromEnd\"]) + \";\" + row[\"name\"]\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    if row[\"Chromosome\"] == \"MT\":\n",
    "        continue\n",
    "    hits = pos2miRbind(row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"])\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_position\"]])\n",
    "        me = me.append([row[\"End_position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\thas 8191  \t miR genes\n",
      "chr2\thas 6877  \t miR genes\n",
      "chr3\thas 5158  \t miR genes\n",
      "chr4\thas 3893  \t miR genes\n",
      "chr5\thas 5148  \t miR genes\n",
      "chr6\thas 4712  \t miR genes\n",
      "chr7\thas 3704  \t miR genes\n",
      "chr8\thas 3484  \t miR genes\n",
      "chr9\thas 2937  \t miR genes\n",
      "chr10\thas 3613  \t miR genes\n",
      "chr11\thas 4471  \t miR genes\n",
      "chr12\thas 5465  \t miR genes\n",
      "chr13\thas 2042  \t miR genes\n",
      "chr14\thas 3097  \t miR genes\n",
      "chr15\thas 4023  \t miR genes\n",
      "chr16\thas 4612  \t miR genes\n",
      "chr17\thas 4854  \t miR genes\n",
      "chr18\thas 2054  \t miR genes\n",
      "chr19\thas 4770  \t miR genes\n",
      "chr20\thas 2722  \t miR genes\n",
      "chr21\thas 950  \t miR genes\n",
      "chr22\thas 1784  \t miR genes\n",
      "chrX\thas 3047  \t miR genes\n",
      "chrY\thas 306  \t miR genes\n",
      "ABCA5 17 67302911 67302912 TCGA-OR-A5JI-01A-11D-A29I-10 hits on polyA_signal;67302907;67302913;535691\n"
     ]
    }
   ],
   "source": [
    "fh = pd.read_table(TABLE3)\n",
    "chrom2fh = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2fh[chr] = fh[fh[\"chrom\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d  \\t miR genes\" % (chr, len(chrom2fh[chr]))\n",
    "\n",
    "def pos2polyA(chr, start, end):\n",
    "    area = chrom2fh[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"txStart\"] <= start <= row[\"txEnd\"] or \\\n",
    "           row[\"txStart\"] <= end <= row[\"txEnd\"] or \\\n",
    "           (start < row[\"txStart\"] and row[\"txEnd\"] < end):\n",
    "            miR = str(row[\"name2\"]) + \";\" + str(row[\"txStart\"]) + \";\" + str(row[\"txEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    if row[\"Chromosome\"] == \"MT\":\n",
    "        continue\n",
    "    hits = pos2polyA(row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"])\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_position\"]])\n",
    "        me = me.append([row[\"End_position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>mStart</th>\n",
       "      <th>mEnd</th>\n",
       "      <th>rStart</th>\n",
       "      <th>rEnd</th>\n",
       "      <th>Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Tumor Sample Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17</td>\n",
       "      <td>27188561</td>\n",
       "      <td>27188561</td>\n",
       "      <td>27188551</td>\n",
       "      <td>27188636</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-144</td>\n",
       "      <td>TCGA-OR-A5J5-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>111384236</td>\n",
       "      <td>111384236</td>\n",
       "      <td>111384164</td>\n",
       "      <td>111384240</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-34c</td>\n",
       "      <td>TCGA-OR-A5J8-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X</td>\n",
       "      <td>146312342</td>\n",
       "      <td>146312342</td>\n",
       "      <td>146312238</td>\n",
       "      <td>146312361</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-506</td>\n",
       "      <td>TCGA-OR-A5JB-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>54210730</td>\n",
       "      <td>54210730</td>\n",
       "      <td>54210707</td>\n",
       "      <td>54210793</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-520c</td>\n",
       "      <td>TCGA-OR-A5JC-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>54210730</td>\n",
       "      <td>54210730</td>\n",
       "      <td>54210722</td>\n",
       "      <td>54210743</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hsa-miR-520c-5p</td>\n",
       "      <td>TCGA-OR-A5JC-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>20</td>\n",
       "      <td>61809858</td>\n",
       "      <td>61809858</td>\n",
       "      <td>61809852</td>\n",
       "      <td>61809938</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-124-3</td>\n",
       "      <td>TCGA-OR-A5JW-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>11</td>\n",
       "      <td>64658676</td>\n",
       "      <td>64658676</td>\n",
       "      <td>64658609</td>\n",
       "      <td>64658718</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-192</td>\n",
       "      <td>TCGA-OR-A5K4-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>11</td>\n",
       "      <td>64658676</td>\n",
       "      <td>64658676</td>\n",
       "      <td>64658675</td>\n",
       "      <td>64658695</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hsa-miR-192-5p</td>\n",
       "      <td>TCGA-OR-A5K4-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>X</td>\n",
       "      <td>146341225</td>\n",
       "      <td>146341225</td>\n",
       "      <td>146341170</td>\n",
       "      <td>146341244</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-509-3</td>\n",
       "      <td>TCGA-OR-A5KB-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>X</td>\n",
       "      <td>146341225</td>\n",
       "      <td>146341225</td>\n",
       "      <td>146341214</td>\n",
       "      <td>146341235</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hsa-miR-509-3-5p</td>\n",
       "      <td>TCGA-OR-A5KB-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>14</td>\n",
       "      <td>101351118</td>\n",
       "      <td>101351118</td>\n",
       "      <td>101351039</td>\n",
       "      <td>101351120</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-136</td>\n",
       "      <td>TCGA-OR-A5KB-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>X</td>\n",
       "      <td>146312360</td>\n",
       "      <td>146312360</td>\n",
       "      <td>146312238</td>\n",
       "      <td>146312361</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-506</td>\n",
       "      <td>TCGA-OR-A5KB-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>X</td>\n",
       "      <td>145078780</td>\n",
       "      <td>145078780</td>\n",
       "      <td>145078716</td>\n",
       "      <td>145078792</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-892b</td>\n",
       "      <td>TCGA-OR-A5KB-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>19</td>\n",
       "      <td>54215536</td>\n",
       "      <td>54215536</td>\n",
       "      <td>54215522</td>\n",
       "      <td>54215608</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-517a</td>\n",
       "      <td>TCGA-OR-A5KB-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>19</td>\n",
       "      <td>54215536</td>\n",
       "      <td>54215536</td>\n",
       "      <td>54215536</td>\n",
       "      <td>54215557</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hsa-miR-517-5p</td>\n",
       "      <td>TCGA-OR-A5KB-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>19</td>\n",
       "      <td>54169970</td>\n",
       "      <td>54169970</td>\n",
       "      <td>54169933</td>\n",
       "      <td>54170016</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-512-1</td>\n",
       "      <td>TCGA-OR-A5KT-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>19</td>\n",
       "      <td>54185480</td>\n",
       "      <td>54185480</td>\n",
       "      <td>54185413</td>\n",
       "      <td>54185499</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-520f</td>\n",
       "      <td>TCGA-OR-A5KZ-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>19</td>\n",
       "      <td>54185480</td>\n",
       "      <td>54185480</td>\n",
       "      <td>54185467</td>\n",
       "      <td>54185488</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hsa-miR-520f-3p</td>\n",
       "      <td>TCGA-OR-A5KZ-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>20</td>\n",
       "      <td>57393302</td>\n",
       "      <td>57393302</td>\n",
       "      <td>57393281</td>\n",
       "      <td>57393368</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-298</td>\n",
       "      <td>TCGA-OR-A5LB-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>54172505</td>\n",
       "      <td>54172505</td>\n",
       "      <td>54172411</td>\n",
       "      <td>54172508</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-512-2</td>\n",
       "      <td>TCGA-OR-A5LD-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>X</td>\n",
       "      <td>133675462</td>\n",
       "      <td>133675462</td>\n",
       "      <td>133675371</td>\n",
       "      <td>133675467</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-542</td>\n",
       "      <td>TCGA-OR-A5LI-01A-11D-A30A-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>6</td>\n",
       "      <td>160903734</td>\n",
       "      <td>160903734</td>\n",
       "      <td>160903734</td>\n",
       "      <td>160903741</td>\n",
       "      <td>miR binding site</td>\n",
       "      <td>LPAL2:miR-376/376ab/376b-3p:2</td>\n",
       "      <td>TCGA-OR-A5JC-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>17</td>\n",
       "      <td>67302911</td>\n",
       "      <td>67302912</td>\n",
       "      <td>67302907</td>\n",
       "      <td>67302913</td>\n",
       "      <td>polyA_signal</td>\n",
       "      <td>535691</td>\n",
       "      <td>TCGA-OR-A5JI-01A-11D-A29I-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom     mStart       mEnd     rStart       rEnd  \\\n",
       "0     17   27188561   27188561   27188551   27188636   \n",
       "1     11  111384236  111384236  111384164  111384240   \n",
       "2      X  146312342  146312342  146312238  146312361   \n",
       "3     19   54210730   54210730   54210707   54210793   \n",
       "4     19   54210730   54210730   54210722   54210743   \n",
       "5     20   61809858   61809858   61809852   61809938   \n",
       "6     11   64658676   64658676   64658609   64658718   \n",
       "7     11   64658676   64658676   64658675   64658695   \n",
       "8      X  146341225  146341225  146341170  146341244   \n",
       "9      X  146341225  146341225  146341214  146341235   \n",
       "10    14  101351118  101351118  101351039  101351120   \n",
       "11     X  146312360  146312360  146312238  146312361   \n",
       "12     X  145078780  145078780  145078716  145078792   \n",
       "13    19   54215536   54215536   54215522   54215608   \n",
       "14    19   54215536   54215536   54215536   54215557   \n",
       "15    19   54169970   54169970   54169933   54170016   \n",
       "16    19   54185480   54185480   54185413   54185499   \n",
       "17    19   54185480   54185480   54185467   54185488   \n",
       "18    20   57393302   57393302   57393281   57393368   \n",
       "19    19   54172505   54172505   54172411   54172508   \n",
       "20     X  133675462  133675462  133675371  133675467   \n",
       "21     6  160903734  160903734  160903734  160903741   \n",
       "22    17   67302911   67302912   67302907   67302913   \n",
       "\n",
       "                        Type                             ID  \\\n",
       "0   miRNA_primary_transcript                    hsa-mir-144   \n",
       "1   miRNA_primary_transcript                    hsa-mir-34c   \n",
       "2   miRNA_primary_transcript                    hsa-mir-506   \n",
       "3   miRNA_primary_transcript                   hsa-mir-520c   \n",
       "4                      miRNA                hsa-miR-520c-5p   \n",
       "5   miRNA_primary_transcript                  hsa-mir-124-3   \n",
       "6   miRNA_primary_transcript                    hsa-mir-192   \n",
       "7                      miRNA                 hsa-miR-192-5p   \n",
       "8   miRNA_primary_transcript                  hsa-mir-509-3   \n",
       "9                      miRNA               hsa-miR-509-3-5p   \n",
       "10  miRNA_primary_transcript                    hsa-mir-136   \n",
       "11  miRNA_primary_transcript                    hsa-mir-506   \n",
       "12  miRNA_primary_transcript                   hsa-mir-892b   \n",
       "13  miRNA_primary_transcript                   hsa-mir-517a   \n",
       "14                     miRNA                 hsa-miR-517-5p   \n",
       "15  miRNA_primary_transcript                  hsa-mir-512-1   \n",
       "16  miRNA_primary_transcript                   hsa-mir-520f   \n",
       "17                     miRNA                hsa-miR-520f-3p   \n",
       "18  miRNA_primary_transcript                    hsa-mir-298   \n",
       "19  miRNA_primary_transcript                  hsa-mir-512-2   \n",
       "20  miRNA_primary_transcript                    hsa-mir-542   \n",
       "21          miR binding site  LPAL2:miR-376/376ab/376b-3p:2   \n",
       "22              polyA_signal                         535691   \n",
       "\n",
       "            Tumor Sample Barcode  \n",
       "0   TCGA-OR-A5J5-01A-11D-A29I-10  \n",
       "1   TCGA-OR-A5J8-01A-11D-A29I-10  \n",
       "2   TCGA-OR-A5JB-01A-11D-A29I-10  \n",
       "3   TCGA-OR-A5JC-01A-11D-A29I-10  \n",
       "4   TCGA-OR-A5JC-01A-11D-A29I-10  \n",
       "5   TCGA-OR-A5JW-01A-11D-A29I-10  \n",
       "6   TCGA-OR-A5K4-01A-11D-A29I-10  \n",
       "7   TCGA-OR-A5K4-01A-11D-A29I-10  \n",
       "8   TCGA-OR-A5KB-01A-11D-A30A-10  \n",
       "9   TCGA-OR-A5KB-01A-11D-A30A-10  \n",
       "10  TCGA-OR-A5KB-01A-11D-A30A-10  \n",
       "11  TCGA-OR-A5KB-01A-11D-A30A-10  \n",
       "12  TCGA-OR-A5KB-01A-11D-A30A-10  \n",
       "13  TCGA-OR-A5KB-01A-11D-A30A-10  \n",
       "14  TCGA-OR-A5KB-01A-11D-A30A-10  \n",
       "15  TCGA-OR-A5KT-01A-11D-A29I-10  \n",
       "16  TCGA-OR-A5KZ-01A-11D-A29I-10  \n",
       "17  TCGA-OR-A5KZ-01A-11D-A29I-10  \n",
       "18  TCGA-OR-A5LB-01A-11D-A29I-10  \n",
       "19  TCGA-OR-A5LD-01A-11D-A29I-10  \n",
       "20  TCGA-OR-A5LI-01A-11D-A30A-10  \n",
       "21  TCGA-OR-A5JC-01A-11D-A29I-10  \n",
       "22  TCGA-OR-A5JI-01A-11D-A29I-10  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.rename(columns={0:\"Chrom\"}, inplace =True)\n",
    "ms.rename(columns={0:\"mStart\"}, inplace=True)\n",
    "me.rename(columns={0:\"mEnd\"}, inplace=True)\n",
    "rs.rename(columns={0:\"rStart\"}, inplace=True)\n",
    "re.rename(columns={0:\"rEnd\"}, inplace=True)\n",
    "t.rename(columns={0:\"Type\"}, inplace=True)\n",
    "i.rename(columns={0:\"ID\"}, inplace=True)\n",
    "s.rename(columns={0:\"Tumor Sample Barcode\"}, inplace=True)\n",
    "final = pd.concat([c,ms,me,rs,re,t,i,s], axis=1)\n",
    "final = final.reset_index(drop=True)\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final.to_excel(\"ACChits.xls\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "NT_113923\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n"
     ]
    }
   ],
   "source": [
    "FILE = \"C:/Users/jtso/My Documents/LAML1.txt\"\n",
    "\n",
    "maf = pd.read_table(FILE, usecols=[\"Hugo_Symbol\", \"Chromosome\", \"Start_Position\", \"End_Position\", \"Tumor_Sample_Barcode\"])\n",
    "\n",
    "FILE = \"../hw6/hsa.gff3\"\n",
    "TABLE2 = \"C:/Users/jtso/Downloads/targetScanS.gz\"  # miRNA binding sites\n",
    "TABLE3 = \"C:/Users/jtso/Downloads/wgEncodeGencodePolyaV19 (1).gz\"  # polyA sites\n",
    "\n",
    "df = pd.read_table(FILE, comment=\"#\", usecols=[\"Chromosome\", \"Start\", \"End\", \"Type\", \"miR Name\"])\n",
    "\n",
    "chr2df = {}\n",
    "for chr in [str(i) for i in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chr2df[chr] = df[df[\"Chromosome\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miR(chr, start, end):\n",
    "    chunk = chr2df[str(chr)]  # shrink the search space to the given chr\n",
    "    out = []\n",
    "    for idx, row in chunk.iterrows():\n",
    "        if row[\"Start\"] <= start <= row[\"End\"] or \\\n",
    "           row[\"Start\"] <=  end  <= row[\"End\"] or \\\n",
    "           (start < row[\"Start\"] and row[\"End\"] < end):\n",
    "            # get the name (either hsa-mir-### or hsa-miR-###)\n",
    "            miR = str(row[\"Type\"]) + \";\" + str(row[\"Start\"]) + \";\" + str(row[\"End\"]) + \";\" + row[\"miR Name\"].split(\"Name=\")[1].split(\";\")[0]\n",
    "            out.append(miR)\n",
    "    return out\n",
    "\n",
    "c = pd.DataFrame()\n",
    "ms = pd.DataFrame()\n",
    "me = pd.DataFrame()\n",
    "rs = pd.DataFrame()\n",
    "re = pd.DataFrame()\n",
    "t = pd.DataFrame()\n",
    "i = pd.DataFrame()\n",
    "s = pd.DataFrame()\n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "NT_113923\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n"
     ]
    }
   ],
   "source": [
    "hi = pd.read_table(TABLE2)\n",
    "\n",
    "chrom2hi = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2hi[chr] = hi[hi[\"chrom\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miRbind(chr, start, end):\n",
    "    area = chrom2hi[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"chromStart\"] <= start <= row[\"chromEnd\"] or \\\n",
    "           row[\"chromStart\"] <= end <= row[\"chromEnd\"] or \\\n",
    "           (start < row[\"chromStart\"] and row[\"chromEnd\"] < end):\n",
    "            miR = \"miR binding site\" + \";\" + str(row[\"chromStart\"]) + \";\" + str(row[\"chromEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\thas 8191  \t miR genes\n",
      "chr2\thas 6877  \t miR genes\n",
      "chr3\thas 5158  \t miR genes\n",
      "chr4\thas 3893  \t miR genes\n",
      "chr5\thas 5148  \t miR genes\n",
      "chr6\thas 4712  \t miR genes\n",
      "chr7\thas 3704  \t miR genes\n",
      "chr8\thas 3484  \t miR genes\n",
      "chr9\thas 2937  \t miR genes\n",
      "chr10\thas 3613  \t miR genes\n",
      "chr11\thas 4471  \t miR genes\n",
      "chr12\thas 5465  \t miR genes\n",
      "chr13\thas 2042  \t miR genes\n",
      "chr14\thas 3097  \t miR genes\n",
      "chr15\thas 4023  \t miR genes\n",
      "chr16\thas 4612  \t miR genes\n",
      "chr17\thas 4854  \t miR genes\n",
      "chr18\thas 2054  \t miR genes\n",
      "chr19\thas 4770  \t miR genes\n",
      "chr20\thas 2722  \t miR genes\n",
      "chr21\thas 950  \t miR genes\n",
      "chr22\thas 1784  \t miR genes\n",
      "chrX\thas 3047  \t miR genes\n",
      "chrY\thas 306  \t miR genes\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "NT_113923\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n"
     ]
    }
   ],
   "source": [
    "fh = pd.read_table(TABLE3)\n",
    "chrom2fh = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2fh[chr] = fh[fh[\"chrom\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d  \\t miR genes\" % (chr, len(chrom2fh[chr]))\n",
    "\n",
    "def pos2polyA(chr, start, end):\n",
    "    area = chrom2fh[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"txStart\"] <= start <= row[\"txEnd\"] or \\\n",
    "           row[\"txStart\"] <= end <= row[\"txEnd\"] or \\\n",
    "           (start < row[\"txStart\"] and row[\"txEnd\"] < end):\n",
    "            miR = str(row[\"name2\"]) + \";\" + str(row[\"txStart\"]) + \";\" + str(row[\"txEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.rename(columns={0:\"Chrom\"}, inplace =True)\n",
    "ms.rename(columns={0:\"mStart\"}, inplace=True)\n",
    "me.rename(columns={0:\"mEnd\"}, inplace=True)\n",
    "rs.rename(columns={0:\"rStart\"}, inplace=True)\n",
    "re.rename(columns={0:\"rEnd\"}, inplace=True)\n",
    "t.rename(columns={0:\"Type\"}, inplace=True)\n",
    "i.rename(columns={0:\"ID\"}, inplace=True)\n",
    "s.rename(columns={0:\"Tumor Sample Barcode\"}, inplace=True)\n",
    "final = pd.concat([c,ms,me,rs,re,t,i,s], axis=1)\n",
    "final = final.reset_index(drop=True)\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final.to_excel(\"LAMLhits.xls\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>879822</td>\n",
       "      <td>879830</td>\n",
       "      <td>SAMD11:miR-504</td>\n",
       "      <td>90</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>900599</td>\n",
       "      <td>900606</td>\n",
       "      <td>KLHL17:miR-299/299-3p</td>\n",
       "      <td>26</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>900605</td>\n",
       "      <td>900612</td>\n",
       "      <td>KLHL17:miR-124/506</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>900933</td>\n",
       "      <td>900941</td>\n",
       "      <td>KLHL17:miR-19</td>\n",
       "      <td>82</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>591</td>\n",
       "      <td>chr1</td>\n",
       "      <td>901054</td>\n",
       "      <td>901061</td>\n",
       "      <td>KLHL17:miR-137</td>\n",
       "      <td>14</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #bin chrom  chromStart  chromEnd                   name  score strand\n",
       "0   591  chr1      879822    879830         SAMD11:miR-504     90      +\n",
       "1   591  chr1      900599    900606  KLHL17:miR-299/299-3p     26      +\n",
       "2   591  chr1      900605    900612     KLHL17:miR-124/506      7      +\n",
       "3   591  chr1      900933    900941          KLHL17:miR-19     82      +\n",
       "4   591  chr1      901054    901061         KLHL17:miR-137     14      +"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
