{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KIAA0090</td>\n",
       "      <td>1</td>\n",
       "      <td>19549914</td>\n",
       "      <td>19549914</td>\n",
       "      <td>TCGA-BL-A0C8-01A-11D-A10S-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COL9A2</td>\n",
       "      <td>1</td>\n",
       "      <td>40766912</td>\n",
       "      <td>40766912</td>\n",
       "      <td>TCGA-BL-A0C8-01A-11D-A10S-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SLC6A9</td>\n",
       "      <td>1</td>\n",
       "      <td>44474107</td>\n",
       "      <td>44474107</td>\n",
       "      <td>TCGA-BL-A0C8-01A-11D-A10S-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HSPB11</td>\n",
       "      <td>1</td>\n",
       "      <td>54395807</td>\n",
       "      <td>54395807</td>\n",
       "      <td>TCGA-BL-A0C8-01A-11D-A10S-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ITGA10</td>\n",
       "      <td>1</td>\n",
       "      <td>145534100</td>\n",
       "      <td>145534100</td>\n",
       "      <td>TCGA-BL-A0C8-01A-11D-A10S-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Chromosome  Start_position  End_position  \\\n",
       "0    KIAA0090          1        19549914      19549914   \n",
       "1      COL9A2          1        40766912      40766912   \n",
       "2      SLC6A9          1        44474107      44474107   \n",
       "3      HSPB11          1        54395807      54395807   \n",
       "4      ITGA10          1       145534100     145534100   \n",
       "\n",
       "           Tumor_Sample_Barcode  \n",
       "0  TCGA-BL-A0C8-01A-11D-A10S-08  \n",
       "1  TCGA-BL-A0C8-01A-11D-A10S-08  \n",
       "2  TCGA-BL-A0C8-01A-11D-A10S-08  \n",
       "3  TCGA-BL-A0C8-01A-11D-A10S-08  \n",
       "4  TCGA-BL-A0C8-01A-11D-A10S-08  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pylab import *\n",
    "\n",
    "FILE = \"C:/Users/jtso/My Documents/BLCA1.txt\"\n",
    "\n",
    "maf = pd.read_table(FILE, skiprows = 3, usecols=[\"Hugo_Symbol\", \"Chromosome\", \"Start_position\", \"End_position\", \"Tumor_Sample_Barcode\"])\n",
    "maf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39312\n"
     ]
    }
   ],
   "source": [
    "print len(maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOC642587 1 209605571 209605571 TCGA-BT-A0YX-01A-11D-A10S-08 hits on miRNA_primary_transcript;209605478;209605587;hsa-mir-205\n",
      "HGS 17 79660899 79660899 TCGA-DK-A1AF-01A-11D-A13W-08 hits on miRNA_primary_transcript;79660787;79660899;hsa-mir-6786\n",
      "FOXF2 6 1390560 1390560 TCGA-DK-A3IQ-01A-31D-A20D-08 hits on miRNA_primary_transcript;1390549;1390646;hsa-mir-6720\n",
      "FOXF2 6 1390560 1390560 TCGA-DK-A3IQ-01A-31D-A20D-08 hits on miRNA;1390559;1390580;hsa-miR-6720-3p\n",
      "FOXF2 6 1390569 1390569 TCGA-DK-A3IQ-01A-31D-A20D-08 hits on miRNA_primary_transcript;1390549;1390646;hsa-mir-6720\n",
      "FOXF2 6 1390569 1390569 TCGA-DK-A3IQ-01A-31D-A20D-08 hits on miRNA;1390559;1390580;hsa-miR-6720-3p\n",
      "RTL1 14 101349412 101349412 TCGA-FD-A3SM-01A-11D-A22Z-08 hits on miRNA_primary_transcript;101349316;101349412;hsa-mir-127\n",
      "ASH1L 1 155316194 155316194 TCGA-FD-A3SN-01A-12D-A22Z-08 hits on miRNA_primary_transcript;155316141;155316236;hsa-mir-555\n",
      "MCPH1 8 6357451 6357451 TCGA-BL-A13J-01A-11D-A10S-08 hits on miR binding site;6357451;6357458;ANGPT2:miR-224\n",
      "UBXN1 11 62444099 62444099 TCGA-DK-A3IU-01A-11D-A20D-08 hits on miR binding site;62444099;62444106;LOC51035:miR-124/506\n",
      "ACBD4 17 43220992 43220992 TCGA-FJ-A3ZE-01A-11D-A23M-08 hits on miR binding site;43220987;43220994;ACBD4:miR-24\n",
      "MARK4 19 45805792 45805792 TCGA-GV-A3JX-01A-11D-A20D-08 hits on miR binding site;45805787;45805794;MARK4:miR-873\n",
      "MARK4 19 45805792 45805792 TCGA-GV-A3JX-01A-11D-A20D-08 hits on miR binding site;45805790;45805797;MARK4:miR-214/761\n",
      "EFCAB4A 11 831640 831642 TCGA-DK-A3IV-01A-22D-A21A-08 hits on miR binding site;831640;831647;EFCAB4A:miR-15/16/195/424/497:1\n",
      "chr1\thas 8191  \t miR genes\n",
      "chr2\thas 6877  \t miR genes\n",
      "chr3\thas 5158  \t miR genes\n",
      "chr4\thas 3893  \t miR genes\n",
      "chr5\thas 5148  \t miR genes\n",
      "chr6\thas 4712  \t miR genes\n",
      "chr7\thas 3704  \t miR genes\n",
      "chr8\thas 3484  \t miR genes\n",
      "chr9\thas 2937  \t miR genes\n",
      "chr10\thas 3613  \t miR genes\n",
      "chr11\thas 4471  \t miR genes\n",
      "chr12\thas 5465  \t miR genes\n",
      "chr13\thas 2042  \t miR genes\n",
      "chr14\thas 3097  \t miR genes\n",
      "chr15\thas 4023  \t miR genes\n",
      "chr16\thas 4612  \t miR genes\n",
      "chr17\thas 4854  \t miR genes\n",
      "chr18\thas 2054  \t miR genes\n",
      "chr19\thas 4770  \t miR genes\n",
      "chr20\thas 2722  \t miR genes\n",
      "chr21\thas 950  \t miR genes\n",
      "chr22\thas 1784  \t miR genes\n",
      "chrX\thas 3047  \t miR genes\n",
      "chrY\thas 306  \t miR genes\n"
     ]
    }
   ],
   "source": [
    "FILE = \"../hw6/hsa.gff3\"\n",
    "TABLE2 = \"C:/Users/jtso/Downloads/targetScanS.gz\"  # miRNA binding sites\n",
    "TABLE3 = \"C:/Users/jtso/Downloads/wgEncodeGencodePolyaV19 (1).gz\"  # polyA sites\n",
    "df = pd.read_table(FILE, comment=\"#\", usecols=[\"Chromosome\", \"Start\", \"End\", \"Type\", \"miR Name\"])\n",
    "\n",
    "chr2df = {}\n",
    "for chr in [str(i) for i in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chr2df[chr] = df[df[\"Chromosome\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miR(chr, start, end):\n",
    "    chunk = chr2df[str(chr)]  # shrink the search space to the given chr\n",
    "    out = []\n",
    "    for idx, row in chunk.iterrows():\n",
    "        if row[\"Start\"] <= start <= row[\"End\"] or \\\n",
    "           row[\"Start\"] <=  end  <= row[\"End\"] or \\\n",
    "           (start < row[\"Start\"] and row[\"End\"] < end):\n",
    "            # get the name (either hsa-mir-### or hsa-miR-###)\n",
    "            miR = row[\"Type\"] + \";\" + str(row[\"Start\"]) + \";\" + str(row[\"End\"]) + \";\" + row[\"miR Name\"].split(\"Name=\")[1].split(\";\")[0]\n",
    "            out.append(miR)\n",
    "    return out\n",
    "\n",
    "c = pd.DataFrame()\n",
    "ms = pd.DataFrame()\n",
    "me = pd.DataFrame()\n",
    "rs = pd.DataFrame()\n",
    "re = pd.DataFrame()\n",
    "t = pd.DataFrame()\n",
    "i = pd.DataFrame()\n",
    "s = pd.DataFrame()\n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    if row[\"Chromosome\"] == \"MT\":\n",
    "        continue\n",
    "    hits = pos2miR(row[\"Chromosome\"], (row[\"Start_position\"]), (row[\"End_position\"]))\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_position\"]])\n",
    "        me = me.append([row[\"End_position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1\n",
    "        \n",
    "hi = pd.read_table(TABLE2)\n",
    "\n",
    "chrom2hi = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2hi[chr] = hi[hi[\"chrom\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miRbind(chr, start, end):\n",
    "    area = chrom2hi[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"chromStart\"] <= start <= row[\"chromEnd\"] or \\\n",
    "           row[\"chromStart\"] <= end <= row[\"chromEnd\"] or \\\n",
    "           (start < row[\"chromStart\"] and row[\"chromEnd\"] < end):\n",
    "            miR = \"miR binding site\" + \";\" + str(row[\"chromStart\"]) + \";\" + str(row[\"chromEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    if row[\"Chromosome\"] == \"MT\":\n",
    "        continue\n",
    "    hits = pos2miRbind(row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"])\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_position\"]])\n",
    "        me = me.append([row[\"End_position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1\n",
    "        \n",
    "fh = pd.read_table(TABLE3)\n",
    "chrom2fh = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2fh[chr] = fh[fh[\"chrom\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d  \\t miR genes\" % (chr, len(chrom2fh[chr]))\n",
    "\n",
    "def pos2polyA(chr, start, end):\n",
    "    area = chrom2fh[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"txStart\"] <= start <= row[\"txEnd\"] or \\\n",
    "           row[\"txStart\"] <= end <= row[\"txEnd\"] or \\\n",
    "           (start < row[\"txStart\"] and row[\"txEnd\"] < end):\n",
    "            miR = str(row[\"name2\"]) + \";\" + str(row[\"txStart\"]) + \";\" + str(row[\"txEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    if row[\"Chromosome\"] == \"MT\":\n",
    "        continue\n",
    "    hits = pos2polyA(row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"])\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_position\"], row[\"End_position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_position\"]])\n",
    "        me = me.append([row[\"End_position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>mStart</th>\n",
       "      <th>mEnd</th>\n",
       "      <th>rStart</th>\n",
       "      <th>rEnd</th>\n",
       "      <th>Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Tumor Sample Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>209605571</td>\n",
       "      <td>209605571</td>\n",
       "      <td>209605478</td>\n",
       "      <td>209605587</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-205</td>\n",
       "      <td>TCGA-BT-A0YX-01A-11D-A10S-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>79660899</td>\n",
       "      <td>79660899</td>\n",
       "      <td>79660787</td>\n",
       "      <td>79660899</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-6786</td>\n",
       "      <td>TCGA-DK-A1AF-01A-11D-A13W-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>1390560</td>\n",
       "      <td>1390560</td>\n",
       "      <td>1390549</td>\n",
       "      <td>1390646</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-6720</td>\n",
       "      <td>TCGA-DK-A3IQ-01A-31D-A20D-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>1390560</td>\n",
       "      <td>1390560</td>\n",
       "      <td>1390559</td>\n",
       "      <td>1390580</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hsa-miR-6720-3p</td>\n",
       "      <td>TCGA-DK-A3IQ-01A-31D-A20D-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>1390569</td>\n",
       "      <td>1390569</td>\n",
       "      <td>1390549</td>\n",
       "      <td>1390646</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-6720</td>\n",
       "      <td>TCGA-DK-A3IQ-01A-31D-A20D-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>1390569</td>\n",
       "      <td>1390569</td>\n",
       "      <td>1390559</td>\n",
       "      <td>1390580</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hsa-miR-6720-3p</td>\n",
       "      <td>TCGA-DK-A3IQ-01A-31D-A20D-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>14</td>\n",
       "      <td>101349412</td>\n",
       "      <td>101349412</td>\n",
       "      <td>101349316</td>\n",
       "      <td>101349412</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-127</td>\n",
       "      <td>TCGA-FD-A3SM-01A-11D-A22Z-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>155316194</td>\n",
       "      <td>155316194</td>\n",
       "      <td>155316141</td>\n",
       "      <td>155316236</td>\n",
       "      <td>miRNA_primary_transcript</td>\n",
       "      <td>hsa-mir-555</td>\n",
       "      <td>TCGA-FD-A3SN-01A-12D-A22Z-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>6357451</td>\n",
       "      <td>6357451</td>\n",
       "      <td>6357451</td>\n",
       "      <td>6357458</td>\n",
       "      <td>miR binding site</td>\n",
       "      <td>ANGPT2:miR-224</td>\n",
       "      <td>TCGA-BL-A13J-01A-11D-A10S-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>11</td>\n",
       "      <td>62444099</td>\n",
       "      <td>62444099</td>\n",
       "      <td>62444099</td>\n",
       "      <td>62444106</td>\n",
       "      <td>miR binding site</td>\n",
       "      <td>LOC51035:miR-124/506</td>\n",
       "      <td>TCGA-DK-A3IU-01A-11D-A20D-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>17</td>\n",
       "      <td>43220992</td>\n",
       "      <td>43220992</td>\n",
       "      <td>43220987</td>\n",
       "      <td>43220994</td>\n",
       "      <td>miR binding site</td>\n",
       "      <td>ACBD4:miR-24</td>\n",
       "      <td>TCGA-FJ-A3ZE-01A-11D-A23M-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>19</td>\n",
       "      <td>45805792</td>\n",
       "      <td>45805792</td>\n",
       "      <td>45805787</td>\n",
       "      <td>45805794</td>\n",
       "      <td>miR binding site</td>\n",
       "      <td>MARK4:miR-873</td>\n",
       "      <td>TCGA-GV-A3JX-01A-11D-A20D-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>19</td>\n",
       "      <td>45805792</td>\n",
       "      <td>45805792</td>\n",
       "      <td>45805790</td>\n",
       "      <td>45805797</td>\n",
       "      <td>miR binding site</td>\n",
       "      <td>MARK4:miR-214/761</td>\n",
       "      <td>TCGA-GV-A3JX-01A-11D-A20D-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>11</td>\n",
       "      <td>831640</td>\n",
       "      <td>831642</td>\n",
       "      <td>831640</td>\n",
       "      <td>831647</td>\n",
       "      <td>miR binding site</td>\n",
       "      <td>EFCAB4A:miR-15/16/195/424/497:1</td>\n",
       "      <td>TCGA-DK-A3IV-01A-22D-A21A-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom     mStart       mEnd     rStart       rEnd  \\\n",
       "0      1  209605571  209605571  209605478  209605587   \n",
       "1     17   79660899   79660899   79660787   79660899   \n",
       "2      6    1390560    1390560    1390549    1390646   \n",
       "3      6    1390560    1390560    1390559    1390580   \n",
       "4      6    1390569    1390569    1390549    1390646   \n",
       "5      6    1390569    1390569    1390559    1390580   \n",
       "6     14  101349412  101349412  101349316  101349412   \n",
       "7      1  155316194  155316194  155316141  155316236   \n",
       "8      8    6357451    6357451    6357451    6357458   \n",
       "9     11   62444099   62444099   62444099   62444106   \n",
       "10    17   43220992   43220992   43220987   43220994   \n",
       "11    19   45805792   45805792   45805787   45805794   \n",
       "12    19   45805792   45805792   45805790   45805797   \n",
       "13    11     831640     831642     831640     831647   \n",
       "\n",
       "                        Type                               ID  \\\n",
       "0   miRNA_primary_transcript                      hsa-mir-205   \n",
       "1   miRNA_primary_transcript                     hsa-mir-6786   \n",
       "2   miRNA_primary_transcript                     hsa-mir-6720   \n",
       "3                      miRNA                  hsa-miR-6720-3p   \n",
       "4   miRNA_primary_transcript                     hsa-mir-6720   \n",
       "5                      miRNA                  hsa-miR-6720-3p   \n",
       "6   miRNA_primary_transcript                      hsa-mir-127   \n",
       "7   miRNA_primary_transcript                      hsa-mir-555   \n",
       "8           miR binding site                   ANGPT2:miR-224   \n",
       "9           miR binding site             LOC51035:miR-124/506   \n",
       "10          miR binding site                     ACBD4:miR-24   \n",
       "11          miR binding site                    MARK4:miR-873   \n",
       "12          miR binding site                MARK4:miR-214/761   \n",
       "13          miR binding site  EFCAB4A:miR-15/16/195/424/497:1   \n",
       "\n",
       "            Tumor Sample Barcode  \n",
       "0   TCGA-BT-A0YX-01A-11D-A10S-08  \n",
       "1   TCGA-DK-A1AF-01A-11D-A13W-08  \n",
       "2   TCGA-DK-A3IQ-01A-31D-A20D-08  \n",
       "3   TCGA-DK-A3IQ-01A-31D-A20D-08  \n",
       "4   TCGA-DK-A3IQ-01A-31D-A20D-08  \n",
       "5   TCGA-DK-A3IQ-01A-31D-A20D-08  \n",
       "6   TCGA-FD-A3SM-01A-11D-A22Z-08  \n",
       "7   TCGA-FD-A3SN-01A-12D-A22Z-08  \n",
       "8   TCGA-BL-A13J-01A-11D-A10S-08  \n",
       "9   TCGA-DK-A3IU-01A-11D-A20D-08  \n",
       "10  TCGA-FJ-A3ZE-01A-11D-A23M-08  \n",
       "11  TCGA-GV-A3JX-01A-11D-A20D-08  \n",
       "12  TCGA-GV-A3JX-01A-11D-A20D-08  \n",
       "13  TCGA-DK-A3IV-01A-22D-A21A-08  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.rename(columns={0:\"Chrom\"}, inplace =True)\n",
    "ms.rename(columns={0:\"mStart\"}, inplace=True)\n",
    "me.rename(columns={0:\"mEnd\"}, inplace=True)\n",
    "rs.rename(columns={0:\"rStart\"}, inplace=True)\n",
    "re.rename(columns={0:\"rEnd\"}, inplace=True)\n",
    "t.rename(columns={0:\"Type\"}, inplace=True)\n",
    "i.rename(columns={0:\"ID\"}, inplace=True)\n",
    "s.rename(columns={0:\"Tumor Sample Barcode\"}, inplace=True)\n",
    "final = pd.concat([c,ms,me,rs,re,t,i,s], axis=1)\n",
    "final = final.reset_index(drop=True)\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'final' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-ddb47a66abb0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mfinal\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mto_excel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"BLCAhits.xls\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'final' is not defined"
     ]
    }
   ],
   "source": [
    "final.to_excel(\"BLCAhits.xls\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "FILE = \"C:/Users/jtso/My Documents/LUAD1.txt\"\n",
    "\n",
    "maf = pd.read_table(FILE, usecols=[\"Hugo_Symbol\", \"Chromosome\", \"Start_Position\", \"End_Position\", \"Tumor_Sample_Barcode\"])\n",
    "\n",
    "FILE = \"../hw6/hsa.gff3\"\n",
    "TABLE2 = \"C:/Users/jtso/Downloads/targetScanS.gz\"  # miRNA binding sites\n",
    "TABLE3 = \"C:/Users/jtso/Downloads/wgEncodeGencodePolyaV19 (1).gz\"  # polyA sites\n",
    "\n",
    "df = pd.read_table(FILE, comment=\"#\", usecols=[\"Chromosome\", \"Start\", \"End\", \"Type\", \"miR Name\"])\n",
    "\n",
    "chr2df = {}\n",
    "for chr in [str(i) for i in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chr2df[chr] = df[df[\"Chromosome\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miR(chr, start, end):\n",
    "    chunk = chr2df[str(chr)]  # shrink the search space to the given chr\n",
    "    out = []\n",
    "    for idx, row in chunk.iterrows():\n",
    "        if row[\"Start\"] <= start <= row[\"End\"] or \\\n",
    "           row[\"Start\"] <=  end  <= row[\"End\"] or \\\n",
    "           (start < row[\"Start\"] and row[\"End\"] < end):\n",
    "            # get the name (either hsa-mir-### or hsa-miR-###)\n",
    "            miR = str(row[\"Type\"]) + \";\" + str(row[\"Start\"]) + \";\" + str(row[\"End\"]) + \";\" + row[\"miR Name\"].split(\"Name=\")[1].split(\";\")[0]\n",
    "            out.append(miR)\n",
    "    return out\n",
    "\n",
    "c = pd.DataFrame()\n",
    "ms = pd.DataFrame()\n",
    "me = pd.DataFrame()\n",
    "rs = pd.DataFrame()\n",
    "re = pd.DataFrame()\n",
    "t = pd.DataFrame()\n",
    "i = pd.DataFrame()\n",
    "s = pd.DataFrame()\n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hi = pd.read_table(TABLE2)\n",
    "\n",
    "chrom2hi = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2hi[chr] = hi[hi[\"chrom\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miRbind(chr, start, end):\n",
    "    area = chrom2hi[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"chromStart\"] <= start <= row[\"chromEnd\"] or \\\n",
    "           row[\"chromStart\"] <= end <= row[\"chromEnd\"] or \\\n",
    "           (start < row[\"chromStart\"] and row[\"chromEnd\"] < end):\n",
    "            miR = \"miR binding site\" + \";\" + str(row[\"chromStart\"]) + \";\" + str(row[\"chromEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fh = pd.read_table(TABLE3)\n",
    "chrom2fh = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2fh[chr] = fh[fh[\"chrom\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d  \\t miR genes\" % (chr, len(chrom2fh[chr]))\n",
    "\n",
    "def pos2polyA(chr, start, end):\n",
    "    area = chrom2fh[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"txStart\"] <= start <= row[\"txEnd\"] or \\\n",
    "           row[\"txStart\"] <= end <= row[\"txEnd\"] or \\\n",
    "           (start < row[\"txStart\"] and row[\"txEnd\"] < end):\n",
    "            miR = str(row[\"name2\"]) + \";\" + str(row[\"txStart\"]) + \";\" + str(row[\"txEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "c.rename(columns={0:\"Chrom\"}, inplace =True)\n",
    "ms.rename(columns={0:\"mStart\"}, inplace=True)\n",
    "me.rename(columns={0:\"mEnd\"}, inplace=True)\n",
    "rs.rename(columns={0:\"rStart\"}, inplace=True)\n",
    "re.rename(columns={0:\"rEnd\"}, inplace=True)\n",
    "t.rename(columns={0:\"Type\"}, inplace=True)\n",
    "i.rename(columns={0:\"ID\"}, inplace=True)\n",
    "s.rename(columns={0:\"Tumor Sample Barcode\"}, inplace=True)\n",
    "final = pd.concat([c,ms,me,rs,re,t,i,s], axis=1)\n",
    "final = final.reset_index(drop=True)\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final.to_excel(\"LUADhits.xls\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
