{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>10</td>\n",
       "      <td>52587953</td>\n",
       "      <td>52587953</td>\n",
       "      <td>TCGA-D8-A1J8-01A-11D-A13L-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>10</td>\n",
       "      <td>52595854</td>\n",
       "      <td>52595854</td>\n",
       "      <td>TCGA-BH-A0HP-01A-12D-A099-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>10</td>\n",
       "      <td>52595854</td>\n",
       "      <td>52595854</td>\n",
       "      <td>TCGA-BH-A0HP-01A-12D-A099-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>10</td>\n",
       "      <td>52595937</td>\n",
       "      <td>52595937</td>\n",
       "      <td>TCGA-BH-A18P-01A-11D-A12B-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>10</td>\n",
       "      <td>52596055</td>\n",
       "      <td>52596055</td>\n",
       "      <td>TCGA-AC-A2FB-01A-11D-A17D-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        A1CF         10        52587953      52587953   \n",
       "1        A1CF         10        52595854      52595854   \n",
       "2        A1CF         10        52595854      52595854   \n",
       "3        A1CF         10        52595937      52595937   \n",
       "4        A1CF         10        52596055      52596055   \n",
       "\n",
       "           Tumor_Sample_Barcode  \n",
       "0  TCGA-D8-A1J8-01A-11D-A13L-09  \n",
       "1  TCGA-BH-A0HP-01A-12D-A099-09  \n",
       "2  TCGA-BH-A0HP-01A-12D-A099-09  \n",
       "3  TCGA-BH-A18P-01A-11D-A12B-09  \n",
       "4  TCGA-AC-A2FB-01A-11D-A17D-09  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pylab import *\n",
    "\n",
    "FILE = \"C:/Users/jtso/My Documents/BRCA.txt\"\n",
    "\n",
    "maf = pd.read_table(FILE, skiprows = 1, usecols=[\"Hugo_Symbol\", \"Chromosome\", \"Start_Position\", \"End_Position\", \"Tumor_Sample_Barcode\"])\n",
    "maf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "90490\n"
     ]
    }
   ],
   "source": [
    "print len(maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATF5 19 50436338 50436338 TCGA-E2-A15K-06A-11D-A12Q-09 hits on miRNA_primary_transcript;50436321;50436394;hsa-mir-4751\n",
      "ATF5 19 50436338 50436338 TCGA-E2-A15K-06A-11D-A12Q-09 hits on miRNA;50436330;50436353;hsa-miR-4751\n",
      "ATF5 19 50436338 50436338 TCGA-E2-A15K-06A-11D-A12Q-09 hits on miRNA_primary_transcript;50436321;50436394;hsa-mir-4751\n",
      "ATF5 19 50436338 50436338 TCGA-E2-A15K-06A-11D-A12Q-09 hits on miRNA;50436330;50436353;hsa-miR-4751\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "MT\n",
      "F8A1 X 154115715 154115715 TCGA-AN-A0AK-01A-21W-A019-09 hits on miRNA_primary_transcript;154115635;154115733;hsa-mir-1184-1\n",
      "FOXF2 6 1390570 1390570 TCGA-AO-A03M-01B-11D-A10M-09 hits on miRNA_primary_transcript;1390549;1390646;hsa-mir-6720\n",
      "FOXF2 6 1390570 1390570 TCGA-AO-A03M-01B-11D-A10M-09 hits on miRNA;1390559;1390580;hsa-miR-6720-3p\n",
      "IGLV2-8 22 23165351 23165351 TCGA-AO-A0J2-01A-11W-A050-09 hits on miRNA_primary_transcript;23165270;23165365;hsa-mir-650\n",
      "NPC2 14 74946861 74946861 TCGA-A2-A0YE-01A-11D-A10G-09 hits on miRNA_primary_transcript;74946836;74946907;hsa-mir-4709\n",
      "POLM 7 44112992 44112992 TCGA-AC-A23H-01A-11D-A159-09 hits on miRNA_primary_transcript;44112977;44113032;hsa-mir-6838\n",
      "POLM 7 44112992 44112992 TCGA-AC-A23H-01A-11D-A159-09 hits on miRNA;44112977;44112996;hsa-miR-6838-3p\n",
      "PYROXD2 10 100154999 100154999 TCGA-AC-A23H-01A-11D-A159-09 hits on miRNA_primary_transcript;100154975;100155064;hsa-mir-1287\n",
      "PYROXD2 10 100154999 100154999 TCGA-AC-A23H-01A-11D-A159-09 hits on miRNA;100154989;100155010;hsa-miR-1287-3p\n",
      "RTL1 14 101349406 101349406 TCGA-E2-A10C-01A-21D-A10M-09 hits on miRNA_primary_transcript;101349316;101349412;hsa-mir-127\n",
      "SERF2 15 44085898 44085898 TCGA-AN-A0AJ-01A-11W-A019-09 hits on miRNA_primary_transcript;44085857;44085957;hsa-mir-1282\n",
      "SNORA36B 1 220373935 220373935 TCGA-B6-A0IJ-01A-11W-A050-09 hits on miRNA_primary_transcript;220373880;220373961;hsa-mir-664a\n",
      "SNORA36B 1 220373935 220373935 TCGA-B6-A0IJ-01A-11W-A050-09 hits on miRNA;220373928;220373951;hsa-miR-664a-5p\n",
      "SNORA36B 1 220373954 220373954 TCGA-AC-A23H-01A-11D-A159-09 hits on miRNA_primary_transcript;220373880;220373961;hsa-mir-664a\n"
     ]
    }
   ],
   "source": [
    "FILE = \"../hw6/hsa.gff3\"\n",
    "TABLE2 = \"C:/Users/jtso/Downloads/targetScanS.gz\"  # miRNA binding sites\n",
    "TABLE3 = \"C:/Users/jtso/Downloads/wgEncodeGencodePolyaV19 (1).gz\"  # polyA sites\n",
    "df = pd.read_table(FILE, comment=\"#\", usecols=[\"Chromosome\", \"Start\", \"End\", \"Type\", \"miR Name\"])\n",
    "\n",
    "chr2df = {}\n",
    "for chr in [str(i) for i in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chr2df[chr] = df[df[\"Chromosome\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miR(chr, start, end):\n",
    "    chunk = chr2df[str(chr)]  # shrink the search space to the given chr\n",
    "    out = []\n",
    "    for idx, row in chunk.iterrows():\n",
    "        if row[\"Start\"] <= start <= row[\"End\"] or \\\n",
    "           row[\"Start\"] <=  end  <= row[\"End\"] or \\\n",
    "           (start < row[\"Start\"] and row[\"End\"] < end):\n",
    "            # get the name (either hsa-mir-### or hsa-miR-###)\n",
    "            miR = str(row[\"Type\"]) + \";\" + str(row[\"Start\"]) + \";\" + str(row[\"End\"]) + \";\" + row[\"miR Name\"].split(\"Name=\")[1].split(\";\")[0]\n",
    "            out.append(miR)\n",
    "    return out\n",
    "\n",
    "c = pd.DataFrame()\n",
    "ms = pd.DataFrame()\n",
    "me = pd.DataFrame()\n",
    "rs = pd.DataFrame()\n",
    "re = pd.DataFrame()\n",
    "t = pd.DataFrame()\n",
    "i = pd.DataFrame()\n",
    "s = pd.DataFrame()\n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hi = pd.read_table(TABLE2)\n",
    "\n",
    "chrom2hi = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2hi[chr] = hi[hi[\"chrom\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miRbind(chr, start, end):\n",
    "    area = chrom2hi[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"chromStart\"] <= start <= row[\"chromEnd\"] or \\\n",
    "           row[\"chromStart\"] <= end <= row[\"chromEnd\"] or \\\n",
    "           (start < row[\"chromStart\"] and row[\"chromEnd\"] < end):\n",
    "            miR = \"miR binding site\" + \";\" + str(row[\"chromStart\"]) + \";\" + str(row[\"chromEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fh = pd.read_table(TABLE3)\n",
    "chrom2fh = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2fh[chr] = fh[fh[\"chrom\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d  \\t miR genes\" % (chr, len(chrom2fh[chr]))\n",
    "\n",
    "def pos2polyA(chr, start, end):\n",
    "    area = chrom2fh[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"txStart\"] <= start <= row[\"txEnd\"] or \\\n",
    "           row[\"txStart\"] <= end <= row[\"txEnd\"] or \\\n",
    "           (start < row[\"txStart\"] and row[\"txEnd\"] < end):\n",
    "            miR = str(row[\"name2\"]) + \";\" + str(row[\"txStart\"]) + \";\" + str(row[\"txEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "c.rename(columns={0:\"Chrom\"}, inplace =True)\n",
    "ms.rename(columns={0:\"mStart\"}, inplace=True)\n",
    "me.rename(columns={0:\"mEnd\"}, inplace=True)\n",
    "rs.rename(columns={0:\"rStart\"}, inplace=True)\n",
    "re.rename(columns={0:\"rEnd\"}, inplace=True)\n",
    "t.rename(columns={0:\"Type\"}, inplace=True)\n",
    "i.rename(columns={0:\"ID\"}, inplace=True)\n",
    "s.rename(columns={0:\"Tumor Sample Barcode\"}, inplace=True)\n",
    "final = pd.concat([c,ms,me,rs,re,t,i,s], axis=1)\n",
    "final = final.reset_index(drop=True)\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final.to_excel(\"BRCAhits.xls\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "FILE = \"C:/Users/jtso/My Documents/LUSC1.txt\"\n",
    "\n",
    "maf = pd.read_table(FILE, usecols=[\"Hugo_Symbol\", \"Chromosome\", \"Start_Position\", \"End_Position\", \"Tumor_Sample_Barcode\"])\n",
    "\n",
    "FILE = \"../hw6/hsa.gff3\"\n",
    "TABLE2 = \"C:/Users/jtso/Downloads/targetScanS.gz\"  # miRNA binding sites\n",
    "TABLE3 = \"C:/Users/jtso/Downloads/wgEncodeGencodePolyaV19 (1).gz\"  # polyA sites\n",
    "\n",
    "df = pd.read_table(FILE, comment=\"#\", usecols=[\"Chromosome\", \"Start\", \"End\", \"Type\", \"miR Name\"])\n",
    "\n",
    "chr2df = {}\n",
    "for chr in [str(i) for i in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chr2df[chr] = df[df[\"Chromosome\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miR(chr, start, end):\n",
    "    chunk = chr2df[str(chr)]  # shrink the search space to the given chr\n",
    "    out = []\n",
    "    for idx, row in chunk.iterrows():\n",
    "        if row[\"Start\"] <= start <= row[\"End\"] or \\\n",
    "           row[\"Start\"] <=  end  <= row[\"End\"] or \\\n",
    "           (start < row[\"Start\"] and row[\"End\"] < end):\n",
    "            # get the name (either hsa-mir-### or hsa-miR-###)\n",
    "            miR = str(row[\"Type\"]) + \";\" + str(row[\"Start\"]) + \";\" + str(row[\"End\"]) + \";\" + row[\"miR Name\"].split(\"Name=\")[1].split(\";\")[0]\n",
    "            out.append(miR)\n",
    "    return out\n",
    "\n",
    "c = pd.DataFrame()\n",
    "ms = pd.DataFrame()\n",
    "me = pd.DataFrame()\n",
    "rs = pd.DataFrame()\n",
    "re = pd.DataFrame()\n",
    "t = pd.DataFrame()\n",
    "i = pd.DataFrame()\n",
    "s = pd.DataFrame()\n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hi = pd.read_table(TABLE2)\n",
    "\n",
    "chrom2hi = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2hi[chr] = hi[hi[\"chrom\"] == \"chr\" + chr]\n",
    "    \n",
    "def pos2miRbind(chr, start, end):\n",
    "    area = chrom2hi[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"chromStart\"] <= start <= row[\"chromEnd\"] or \\\n",
    "           row[\"chromStart\"] <= end <= row[\"chromEnd\"] or \\\n",
    "           (start < row[\"chromStart\"] and row[\"chromEnd\"] < end):\n",
    "            miR = \"miR binding site\" + \";\" + str(row[\"chromStart\"]) + \";\" + str(row[\"chromEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fh = pd.read_table(TABLE3)\n",
    "chrom2fh = {}\n",
    "for chr in [str(a) for a in range(1,23)] + [\"X\", \"Y\"]:\n",
    "    chrom2fh[chr] = fh[fh[\"chrom\"] == \"chr\" + chr]\n",
    "    print \"chr%s\\thas %d  \\t miR genes\" % (chr, len(chrom2fh[chr]))\n",
    "\n",
    "def pos2polyA(chr, start, end):\n",
    "    area = chrom2fh[str(chr)]\n",
    "    oc = []\n",
    "    for idx, row in area.iterrows():\n",
    "        if row[\"txStart\"] <= start <= row[\"txEnd\"] or \\\n",
    "           row[\"txStart\"] <= end <= row[\"txEnd\"] or \\\n",
    "           (start < row[\"txStart\"] and row[\"txEnd\"] < end):\n",
    "            miR = str(row[\"name2\"]) + \";\" + str(row[\"txStart\"]) + \";\" + str(row[\"txEnd\"]) + \";\" + str(row[\"name\"])\n",
    "            oc.append(miR)\n",
    "    return oc\n",
    "        \n",
    "from collections import Counter\n",
    "miR2hits = Counter()\n",
    "for idx, row in maf.iterrows():\n",
    "    print idx, '\\r',\n",
    "    try:\n",
    "        hits = pos2miR(row[\"Chromosome\"], (row[\"Start_Position\"]), (row[\"End_Position\"]))\n",
    "    except:\n",
    "        print row[\"Chromosome\"]\n",
    "        continue\n",
    "    for hit in hits:\n",
    "        print row[\"Hugo_Symbol\"], row[\"Chromosome\"], row[\"Start_Position\"], row[\"End_Position\"], row[\"Tumor_Sample_Barcode\"], \"hits on\", hit\n",
    "        c = c.append([row[\"Chromosome\"]])\n",
    "        ms = ms.append([row[\"Start_Position\"]])\n",
    "        me = me.append([row[\"End_Position\"]])\n",
    "        rs = rs.append([hit.split(\";\")[1]])\n",
    "        re = re.append([hit.split(\";\")[2]])\n",
    "        t = t.append([hit.split(\";\")[0]])\n",
    "        i = i.append([hit.split(\";\")[3]])\n",
    "        s = s.append([row[\"Tumor_Sample_Barcode\"]])\n",
    "        miR2hits[hit] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "c.rename(columns={0:\"Chrom\"}, inplace =True)\n",
    "ms.rename(columns={0:\"mStart\"}, inplace=True)\n",
    "me.rename(columns={0:\"mEnd\"}, inplace=True)\n",
    "rs.rename(columns={0:\"rStart\"}, inplace=True)\n",
    "re.rename(columns={0:\"rEnd\"}, inplace=True)\n",
    "t.rename(columns={0:\"Type\"}, inplace=True)\n",
    "i.rename(columns={0:\"ID\"}, inplace=True)\n",
    "s.rename(columns={0:\"Tumor Sample Barcode\"}, inplace=True)\n",
    "final = pd.concat([c,ms,me,rs,re,t,i,s], axis=1)\n",
    "final = final.reset_index(drop=True)\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final.to_excel(\"LUSChits.xls\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
